### PR TITLE
Hierarchical search engine review

### PR DESCRIPTION
**Pull request by drexlerd**
Hi,

Here is my working version of a hierarchical search engine. The idea is to have a tree of hierarchical search engines. The step function passes control flow to low-level search engines. Upon reaching a goal state in the leaf searches, the parent search engines are notified, and global search status is returned to the leaf and then returned to the root search engine through the step function to the search function.

I would kindly ask for short feedback on my implementation. More specifically, design and or performance issues. It is fully working, and instructions on how to run it can be found in search_engines/hierarchical_search_engines/README.MD. Sorry for having so many changes in the pull request. The interesting parts are in the subdirectory search_engines/hierarchical_search_engines. I did not manage to make a pull request containing changes to a running version.

Thanks and cheers,
Dominik

### CODE REVIEW

File: src/search/search_engines/hierarchical_search_engines/iterative_width_search.cc [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        int old_num_states = m_state_registry->size();
+        State succ_state = m_state_registry->get_successor_state(state, op);
+        int new_num_states = m_state_registry->size();
+        bool is_new_state = (new_num_states > old_num_states);
+        if (!is_new_state) {
+            continue;
+        }
+
+        statistics.inc_generated();
+        bool novel = is_novel(succ_state);
```
Comment (drexlerd) [inline src/search/search_engines/hierarchical_search_engines/iterative_width_search.cc:79]: Generalized implementation of NoveltyTable for arbitrary k in N is currently unable to consider operator effects that must occur in a tuple.

**Comment by jendrikseipp** [comment]
I'll go over the code after lunch.

File: src/search/search_engines/hierarchical_search_engines/iterative_width_search.cc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    State state = m_state_registry->lookup_state(id);
+    // std::cout << "Initial state: " << m_propositional_task->compute_dlplan_state(state).str() << std::endl;
+    SearchNode node = m_search_space->get_node(state);
+    node.close();
+    assert(!node.is_dead_end());
+    statistics.inc_expanded();
+
+    /* Goal check in initial state. */
+    if (id == m_initial_state_id) {
+        if (m_goal_test->is_goal(m_state_registry->lookup_state(m_initial_state_id), state)) {
```
Comment (jendrikseipp) [inline src/search/search_engines/hierarchical_search_engines/iterative_width_search.cc:58]: Can be simplified to `if (id == m_initial_state_id && m_goal_test->is_goal(state, state))`.

**Review by jendrikseipp — COMMENTED** [review]
Nice work! I like the design. It's quite modular and encapsulated!

**Comment by jendrikseipp** [comment]
This PR can probably be closed, right?

**Pull request closed without merging**